File: frontend/src/types/story.ts
/** User story extraction lifecycle status — matches backend UserStoryStatus enum. */
export type UserStoryStatus =
  'pending_extraction' | 'extracting' | 'extracted' | 'failed_extraction';

/** All valid UserStoryStatus values in lifecycle order. */
export const USER_STORY_STATUSES: UserStoryStatus[] = [
  'pending_extraction',
  'extracting',
  'extracted',
  'failed_extraction',
];

/** Human-readable labels for each UserStoryStatus (i18n keys). */
export const USER_STORY_STATUS_LABELS: Record<UserStoryStatus, string> = {
  pending_extraction: 'stories.status_pending_extraction',
  extracting: 'stories.status_extracting',
  extracted: 'stories.status_extracted',
  failed_extraction: 'stories.status_failed_extraction',
};

/** Valid UserStoryStatus transitions (forward only, terminal states). */
export const VALID_USER_STORY_TRANSITIONS: Record<UserStoryStatus, UserStoryStatus[]> = {
  pending_extraction: ['extracting'],
  extracting: ['extracted', 'failed_extraction'],
  extracted: [],
  failed_extraction: [],
};

/** Check if a UserStoryStatus transition is valid. */
export function isValidUserStoryTransition(
  current: UserStoryStatus,
  next: UserStoryStatus,
): boolean {
  return VALID_USER_STORY_TRANSITIONS[current]?.includes(next) ?? false;
}

export interface UserStory {
  id: string;
  projectId: string;
  actor: string;
  feature: string;
  benefit: string;
  rawText: string;
  status: UserStoryStatus;
  createdAt: string;
}

/* ── CSV import ── */

/** Why a CSV row was skipped as a duplicate during import. */
export type StoryImportDuplicateReason = 'duplicate' | 'duplicate_in_file';

/** A row skipped because it duplicates an existing story or another row in the upload. */
export interface StoryImportDuplicate {
  /** 1-based line number in the uploaded CSV. */
  line: number;
  reason: StoryImportDuplicateReason;
  /** Present when reason is 'duplicate': the story that already exists in the project. */
  existingStoryId?: string;
  /** Present when reason is 'duplicate_in_file': the earlier line it repeats. */
  firstLine?: number;
}

/** A single validation problem with one CSV row (blocking — the whole import is rejected). */
export interface StoryImportIssue {
  /** 1-based line number in the uploaded CSV. */
  line: number;
  reason: string;
  field?: string;
  length?: number;
  max?: number;
  observed?: number;
  expected?: number;
}

/** Successful import summary (201 response). */
export interface StoryImportReport {
  created: number;
  skipped: number;
  totalRows: number;
  duplicates: StoryImportDuplicate[];
  storyIds: string[];
}

/**
 * Typed import failure, extracted from an ApiRequestError.
 * - `rows`: 422 IMPORT_VALIDATION_FAILED — per-row validation issues; nothing was created.
 * - `file`: 422 IMPORT_FILE_REJECTED or 413 IMPORT_FILE_TOO_LARGE — the file itself was rejected.
 */
export type StoryImportFailure =
  | {
      kind: 'rows';
      totalRows: number;
      errors: StoryImportIssue[];
      duplicates: StoryImportDuplicate[];
    }
  | {
      kind: 'file';
      errorCode: 'IMPORT_FILE_REJECTED' | 'IMPORT_FILE_TOO_LARGE';
      reason?: string;
      size?: number;
      max?: number;
    };

File: backend/src/storico/api/schemas/story.py
"""UserStory Pydantic schemas for Storico API."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from storico.domain.entities.user_story import UserStoryStatus


class CreateUserStoryRequest(BaseModel):
    """Request body for creating a new user story."""

    model_config = ConfigDict(extra="forbid")

    project_id: UUID
    actor: str = Field(..., min_length=1, max_length=100)
    feature: str = Field(..., min_length=1, max_length=300)
    benefit: str = Field(..., min_length=1, max_length=300)
    raw_text: str = Field(..., max_length=2000)


class UpdateUserStoryRequest(BaseModel):
    """Request body for updating an existing user story."""

    model_config = ConfigDict(extra="forbid")

    actor: str | None = Field(None, min_length=1, max_length=100)
    feature: str | None = Field(None, min_length=1, max_length=300)
    benefit: str | None = Field(None, min_length=1, max_length=300)
    raw_text: str | None = Field(None, max_length=2000)


class UserStoryResponse(BaseModel):
    """Response body representing a user story."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    project_id: UUID
    actor: str
    feature: str
    benefit: str
    raw_text: str
    created_at: datetime
    status: UserStoryStatus = UserStoryStatus.PENDING_EXTRACTION


class StoryImportErrorItem(BaseModel):
    """One blocking problem on one row of an imported CSV."""

    line: int
    reason: str
    field: str | None = None
    length: int | None = None
    # Serialises as "max" in JSON; that is the published contract.
    max: int | None = None
    # Populated only for ``field_count_mismatch``: the row's field count vs
    # the header's column count.
    observed: int | None = None
    expected: int | None = None


class StoryImportDuplicateItem(BaseModel):
    """One row of an imported CSV skipped because it duplicates a story."""

    line: int
    reason: str
    existing_story_id: str | None = None
    first_line: int | None = None


class StoryImportResponse(BaseModel):
    """Response body for a CSV story import."""

    created: int
    skipped: int
    total_rows: int
    duplicates: list[StoryImportDuplicateItem] = []
    story_ids: list[UUID] = []
